kernel/libs/aster-bigtcp/src/socket/bound/tcp_conn.rs

```diff
@@ -57,17 +60,101 @@ impl<E: Ext> DerefMut for RawTcpSocketExt<E> {
     }
 }
 
+impl<E: Ext> RawTcpSocketExt<E> {
+    /// Checks if the socket may receive any new data.
+    ///
+    /// This is similar to [`RawTcpSocket::may_recv`]. However, this method checks if there can be
+    /// _new_ data. In other words, if there is already buffered data in the socket,
+    /// [`RawTcpSocket::may_recv`] will always return true since it is possible to receive the
+    /// buffered data, but this method may return false if the peer has closed its sending half (so
+    /// no new data can come in).
+    pub fn may_recv_new(&self) -> bool {
+        // See also the implementation of `RawTcpSocket::may_recv`.
+        match self.state() {
+            State::Established => true,
+            // Our sending half is closed, but the peer's sending half is still active.
+            State::FinWait1 | State::FinWait2 => true,
+            _ => false,
+        }
+    }
+
+    /// Checks if the socket is closing.
+    ///
+    /// More specifically, we say a socket is closing if and only if it has sent its FIN packet but
+    /// is still waiting for an ACK packet from the peer to acknowledge the FIN it sent.
+    pub fn is_closing(&self) -> bool {
+        let state = self.state();
+        state == State::FinWait1 || state == State::Closing || state == State::LastAck
+    }
+
+    /// Returns whether the receiving half of the socket is shut down.
+    ///
+    /// This method will return true if and only if [`TcpConnection::shut_recv`] or
+    /// [`TcpConnection::close`] is called.
+    pub fn is_recv_shut(&self) -> bool {
+        self.is_recv_shut
+    }
+
+    /// Checks if the socket is closed by a RST packet.
+    ///
+    /// Note that the flag is automatically cleared when it is read by
+    /// [`TcpConnection::clear_rst_closed`], [`TcpConnection::send`], or [`TcpConnection::recv`].
+    pub fn is_rst_closed(&self) -> bool {
+        self.is_rst_closed
+    }
+}
+
 define_boolean_value!(
     /// Whether the TCP connection became dead.
     TcpConnBecameDead
 );
 
 impl<E: Ext> RawTcpSocketExt<E> {
-    fn on_new_state(
+    /// Checks the TCP state for additional events and whether the connection is dead.
+    fn check_state(
         &mut self,
         this: &Arc<TcpConnectionBg<E>>,
+        old_state: State,
+        old_recv_queue: usize,
+        is_rst: bool,
     ) -> (SocketEvents, TcpConnBecameDead) {
-        if self.may_send() && !self.has_connected {
+        let became_dead = if self.state() != State::Established {
+            // After the connection is closed by the user, no new data can be read, and such unread
+            // data will immediately cause the connection to be reset.
+            // Note that "closed" here means that either (1) `close()` or (2) both `shut_send` and
+            // `shut_recv` are called. In the latter case, there may be some buffered data.
+            if self.is_recv_shut
+                // These are states where the sending half is closed but new data can come in.
```

Comment:
If only `shut_recv` is called (without calling `shut_send`), should new data (the data comes in after `shut_recv`) cause RST generated?

Reply:
No. This is covered in existing regression tests:

https://github.com/asterinas/asterinas/blob/efdc43614e8b0dcc7c77c4cb9bf581f810a24986/test/apps/network/tcp_poll.c#L129-L137